nn.hpp

```diff
@@ -227,8 +236,46 @@ nn<const T*> nn_addr(const T & object) {
     return nn<const T*>(i_promise_i_checked_for_null, &object);
 }
 
+/* Non-nullable equivalents of shared_ptr's specialized casting functions.
+ * These convert through a shared_ptr since nn<shared_ptr<T>> lacks the ref-count-sharing cast
+ * constructor, but thanks to moves there shouldn't be any significant extra cost. */
+template <typename T, typename U>
+nn_shared_ptr<T> nn_static_pointer_cast(const nn_shared_ptr<U> & org_ptr) {
+    auto raw_ptr = static_cast<typename nn_shared_ptr<T>::element_type *>(org_ptr.get());
+    std::shared_ptr<T> nullable_ptr(org_ptr.as_nullable(), raw_ptr);
+    return nn_shared_ptr<T>(i_promise_i_checked_for_null, std::move(nullable_ptr));
+}
+
+template <typename T, typename U>
+std::shared_ptr<T> nn_dynamic_pointer_cast(const nn_shared_ptr<U> & org_ptr) {
+    auto raw_ptr = dynamic_cast<typename std::shared_ptr<T>::element_type *>(org_ptr.get());
+    if (!raw_ptr) {
+        return nullptr;
+    } else {
+        return std::shared_ptr<T>(org_ptr.as_nullable(), raw_ptr);
+    }
+}
+
+template <typename T, typename U>
+nn_shared_ptr<T> nn_const_pointer_cast(const nn_shared_ptr<U> & org_ptr) {
+    auto raw_ptr = const_cast<typename nn_shared_ptr<T>::element_type *>(org_ptr.get());
+    std::shared_ptr<T> nullable_ptr(org_ptr.as_nullable(), raw_ptr);
+    return nn_shared_ptr<T>(i_promise_i_checked_for_null, std::move(nullable_ptr));
+}
+
 } } /* end namespace dropbox::oxygen */
 
+namespace std {
+    template <typename T>
+    struct hash<::dropbox::oxygen::nn<T>> {
+        using argument_type = ::dropbox::oxygen::nn<T>;
+        using result_type = size_t;
+        size_t operator()(const argument_type & obj) const {
```

Comment:
Can do, though FYI git blame tells me this is your code I'm pushing. :)

Result-type is a bit weird, since it's declared in the standard, but also defined to always be size_t.  If it were really variable, I'd say the one here should really be defined based on std::hash<T>::result_type;